### DISCUSSION TITLE
Python/PyVISA Noob  considering writing new instruments

### DISCUSSION THREAD

**Discussion by dazz100** [pymeasure/pymeasure#1011] [General]
Hi
I have been programming unprofessionally for a long time, but I am new to Python and gpib.
I have 4 instruments, only one of which is already  in pymeasure.
I am considering writing instruments for pymeasure, but given my current level of knowledge, I am likely to make a bit of a mess of it.  

Would there be someone patient and willing to critique my code if I go down this path?

**Comment by BenediktBurger** [comment]
We're here for helping you writing good coded instrument drivers.

You may start a pull request as early as possible such that we can help you.

You might look into our "adding instruments" guide. Whenever a question arises, don't hesitate to ask.

**Reply by dazz100** [reply]
Hi
I am getting close to wanting to start writing new old instruments including an HP 3721A Correlator and a Electrometrics EMC-30 EMC receiver.  Right now I am modifying the firmware for the Xyphro Usb-GPIB adapter.  I need to test the firmware.  I want to use a HP 3478A Multimeter as the test instrument.    I see that the HP 3478A is already listed in PyMeasure.  What I am really looking for is an app that lets me control the HP 3478A from the keyboard/GUI.  Where would I look to find that?

**Reply by BenediktBurger** [reply]
With the keyboard in an command line interface, you can just open any python interpreter and import the device from pymeasure:
~~~python
from pymeasure.instruments.hp import HP3478A
inst = HP3478A(resourcestring)
~~~

Here `resourcestring` is the string to the VISA resource (com port or similar), where the device is connected.
This allows to use the device, reading values and setting settings.

**Reply by dazz100** [reply]
Hi
That nugget of info is really helpful.  I haven't read all of the documentation but I didn't find info about the command line interface used like you describe.

Thanks.

**Reply by BenediktBurger** [reply]
There is no special "pymeasure command line interface".
Instead, you can use pymeasure instruments in any REPL environment, be it just plain python in the console, or an Ipython console.

In such a console you write the commands, you would have written into a script and execute them one by one.
There you can then send single commands to your device as needed.

Does this help you?
Do you need anything else?

**Reply by dazz100** [reply]
Yes that helps a lot.  
Up until now I have been editing/running/editing/running  ....  scripts.  
Running console commands will be easier for experimentation.
Thanks. Nothing else needed until my next question  ☺
